db/columns/operations/infer_types.py

```diff
@@ -15,6 +15,7 @@
 MAX_INFERENCE_DAG_DEPTH = 100
 
 TYPE_INFERENCE_DAG = {
+    base.PostgresType.TEXT.value: [],
```

Comment:
The way this is set up, you'll never infer a column to be any type except `TEXT`, if it starts as `TEXT`. You need to put `TEXT` upstream of all other types (i.e., where `String` is now).

You need to change the reverse type map function to avoid an infinite loop.

Reply:
> The way this is set up, you'll never infer a column to be any type except TEXT, if it starts as TEXT.

@mathemancer I'm surprised that this didn't trigger a test failure.

Reply:
Okay. So i need to add `base.PostgresType.TEXT.value` as first value in `base.STRING` and then change the `reverse_type_function()` to avoid an infinite loop.
Please confirm it

Reply:
Can you please elaborate what `_get_reverse_type_map()` function is doing?

Reply:
The way the type inference algorithm works is:
 1. reflect a column.
 2. get its type.
 3. map that type to a string representation through `_get_reverse_type_map`.
 4. Look in the `TYPE_INFERENCE_DAG` dictionary for that string as a key
 5. Get the list value associated with that key.
 6. Attempt to cast the column to the type represented by each list value in turn.
 7. If all fail, you're done (leave the column as-is) if you succeed, go to (1).

@dmos62 Actually, there is a test that checks this, and it's failing, but for the wrong reasons. :hankey: I interpreted this failure incorrectly in previous comments. It's actually failing just because it wasn't updated to the new regime where the default column type is `TEXT`.

We treat all string-like types as the same, except for the extra-restrictive `CHAR`. This is what the little dictionary update in `_get_reverse_type_map` is for. It maps all those types to the same string representation.

@vrutik2809 You should try to accomplish the following:

If a column is of type `TEXT`, in the database, it should have every type as a downstream option. Columns of type `TEXT` in the database will have an SQLAlchemy type of either `sqlalchemy.TEXT` or `sqlalchemy.Text`. (The former when reflected, and the latter when freshly created in the Python layer). To represent this, I prefer replacing the `base.STRING` key with `base.PostgresType.TEXT.value` in the `TYPE_INFERENCE_DAG`, and then modifying the `_get_reverse_type_map` function to point all of the python types `Text`, `TEXT`, and `VARCHAR` (as imported in that file) to `base.PostgresType.TEXT.value` in the reverse type map.

Reply:
Okay. I will look into that.
Thank you